### DISCUSSION TITLE
New image format: FLIF

### DISCUSSION THREAD

**Discussion by h2non** [libvips/libvips#1998] [Ideas]
I create the issue just to share a relatively new image format with very good compression rates: http://flif.info

Perhaps `libvips` could be one of the first libraries supporting it in a near future based on the industry adoption.

Otherwise it's definitely interesting and worth seeing.

**Comment by jcupitt** [comment]
Sounds interesting, though it seems their library is not finalized yet, so it may be too early still. I'll have a quick look.

**Comment by jcupitt** [comment]
I downloaded the current release, v0.1 alpha, but it seems they don't have an API yet, just an executable that can encode and decode. Let's revisit this when they have a libflif and a documented interface.

**Comment by felixbuenemann** [comment]
When I last looked at it the FLIF encoder was prohibitively slow…

**Comment by florentb** [comment]
It seems libflif is out with documentation

**Comment by lovell** [comment]
Looks like the FLIF developers are just starting to deal with reports of security vulnerabilities, so not quite there yet :(

https://github.com/FLIF-hub/FLIF/issues/277
https://github.com/FLIF-hub/FLIF/issues/304

**Comment by jcupitt** [comment]
I had a quick hack:

https://github.com/jcupitt/libvips/tree/add-libflif

https://github.com/jcupitt/libvips/blob/add-libflif/libvips/foreign/flifload.c
- [x] add `flifsave`
- [x] add buffer versions
- [ ] look into supporting header load, though the API makes this difficult
- [ ] look into supporting metadata -- it looks like they have xmp, ipct, icc etc. 
- [ ] needs a `flif.pc` for `configure`
- [x] add tests
- [ ] check docs

It looks like it's too slow to be practical at the moment, and of course it's got a few security issues. It can sit on a branch for a while.

**Comment by jcupitt** [comment]
Before I forget, from the point of view of libvips, libflif has these issues:
1. No `.pc` file makes it hard to compile against.
2. There's a thing to read the image header, but it only works from a memory buffer, it won't work from a file. The header can contain a lot of metadata (icc, xmp, ipct etc etc) and there's no way to tell how much of the file you need to read into memory to get the complete header, so you have to read everything. In other words, you can't read the image header without reading the whole file into memory.  
3. A tile-based version of FLIF will be necessary for larger files, but I see this is in their plans. 
4. I don't see how to save anything other than RGBA ... the command-line interface that flif ship seems to use some extra code that's not in the library.

**Comment by jcupitt** [comment]
`flifload` and `flifsave` are working:

```
$ time vips flifsave k2.jpg k2.flif --effort 60
real    0m18.834s
user    0m18.844s
sys 0m0.040s
```

Ouch! I'll add buffer versions too, then set this aside for now.

**Comment by jcupitt** [comment]
OK, buffer load and save added, tests added. I'll leave this for now.

**Comment by jcupitt** [comment]
It's not clear when FLIF will be ready for wider use, so I'll take the 8.5 milestone off.

**Comment by benepo** [comment]
subscribing

**Comment by benepo** [comment]
Using https://github.com/jcupitt/libvips/tree/add-libflif I can convert from png to flif:

Sample image: https://upload.wikimedia.org/wikipedia/commons/7/70/Example.png

```
./vips flifsave Example.png Example.flif
memory: high-water mark 143.46 KB
```

But when I try to convert from **fits** to flif it throws ```Segmentation fault (core dumped)```

Test file http://chandra.si.edu/photo/2009/crab/fits/crab.fits from http://chandra.si.edu/photo/openFITS/crab.html

```gdb``` says:
```
GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.04) 7.11.1
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from vips...done.
[New LWP 20333]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `./vips flifsave crab.fits crab.flif'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f4e0f8daf9c in ?? () from /usr/local/lib/libflif.so.0
```

**Comment by jcupitt** [comment]
Oh, interesting, thanks for testing this. I'll have a look.

**Comment by jcupitt** [comment]
I've updated the add-libflif branch so it's now based on libvips git master, and updated my copy of libflif to master as well. I had to disable the flifsave test since I think libflif has changed how it encodes alpha since 0.1.

The crash seems to be related to it being a 16-bit FITS image. I see this backtrace:

```
#0  0x00007ffff3b51fb7 in TransformPaletteC<FileIO>::process(ColorRanges const*, std::vector<Image, std::allocator<Image> > const&) ()
   from /home/john/GIT/FLIF/src/libflif.so.0
#1  0x00007ffff3b4fd67 in bool flif_encode<FileIO>(FileIO&, std::vector<Image, std::allocator<Image> >&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, flif_options&) ()
   from /home/john/GIT/FLIF/src/libflif.so.0
#2  0x00007ffff3b72b97 in FLIF_ENCODER::encode_file(char const*) ()
   from /home/john/GIT/FLIF/src/libflif.so.0
#3  0x00007ffff3b62ed9 in flif_encoder_encode_file ()
   from /home/john/GIT/FLIF/src/libflif.so.0
#4  0x00007ffff78da595 in vips_foreign_save_flif_file_build (
    object=0x555555883840) at flifsave.c:473
```

I'll do a bit more digging. I don't think I experimented much with 16 bit flif write.

I noticed in another issue you were interested in online viewing of gigapixel astronomical images (I think). I guess you've looked at DeepZoom with PNG tiles? That would support lossless and 16-bit, which might be enough. There's a lot of existing infrastructure around that format as well.

vips includes a fast DeepZoom writer which can make these pyramids, for example:

```
$ time vips dzsave crab.fits x --suffix .png
real	0m0.142s
user	0m0.180s
sys	0m0.040s
```

Notes on usage of `dzsave` here:

http://jcupitt.github.io/libvips/API/current/Making-image-pyramids.md.html

**Comment by jcupitt** [comment]
Ah, it's not that it's a 16-bit fits, it's that it has bzero / bscale set, so cfitsio gives vips a float image. This then causes horrible breakage inside flifsave.

Try:

```
$ vips cast crab.fits x.flif uchar
$ vips cast crab.fits x.flif ushort
```

And it'll save as 8 or 16 bit successfully. I'll fix the format casting logic in flifsave.

**Comment by benepo** [comment]
Thanks @jcupitt !  I'll check out the new version and also the DeepZoom option for our use case.

_Update regarding DeepZoom:_ pre-generating the full tile pyramid would require significant additional storage costs.  Instead, we hope to leverage a solution like [iipsrv](https://github.com/ruven/iipsrv/blob/4a4cd19808e607c80eeec4ace89a90466612e6ee/README#L20) for dynamic on-the-fly tile generation and caching.

**Comment by benepo** [comment]
Worked like a charm!  Any guidance on when I should use uchar vs ushort?  Both commands produced an identical FLIF image.  Also, any thoughts on adding FLIF support in your next official release?

**Comment by jcupitt** [comment]
You're right, pre-rendering the pyr will need more storage (and CPU!), but typically "only" 30%, which isn't that large, in my opinion.

1 + 1/4 + 1/16 + 1/64 + ... ~= 1.30

The real cost can be much higher. If you are storing 1kb tiles, and files are stored in 4kb pages, you are wasting 75% of your disc. However, dzsave supports zip output:

```
vips dzsave crab.fits x.zip --suffix .png
```

Now it'll write a single file: an uncompressed zip containing all the tiles in the usual directory structure.

You can read this from a javascript client with an http resume request to seek to the end and extract the directory structure, then a single http resume to pull out each tile. 

The pro would be low latency for clients, low load on the server, and the ability to use a cheap, standard, easy to maintain, unmodified web server. The con would be somewhat higher storage costs, and high upfront CPU.

**Comment by jcupitt** [comment]
I'm trying to fix it so you can just copy the files, there's some problem in the image type conversion code, stupid thing. 

For now, just use `uchar` everywhere I think.

**Comment by jcupitt** [comment]
Oh argh, I give up. I have two u16 PNGs now, one works, one segvs, they only differ in their pixel values. 

The segv one will encode with `flif -e`, but I can't work out or duplicate what options the flif program is setting, because it doesn't use the library API, it manipulates all the internal classes directly. 

Let's leave using the cast as a workaround for now, and revisit when libflif is more mature.

**Comment by jcupitt** [comment]
Last post! You asked about merging libflif support into the main branch of libvips. I like flif, but it all feels rather immature at the moment. Maybe when Debian have a package.
